### PR TITLE
support nodeSelector in kubefed init

### LINKED ISSUE #50716: Enable `kubefed init` support nodeSelector
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line: 
>
> /kind bug
/kind feature

**What happened**:

The `kubefed init` will install federation control panel on worker node randomly, sometimes, the user have some requirement want to use `nodeSelector` to have some it run on some dedicate node.

The `kubefed init` cannot enable users specify `nodeSelector` for now, but only options as following:
```
      --api-server-advertise-address string      Preferred address to advertise api server nodeport service. Valid only if 'api-server-service-type=NodePort'.
      --api-server-service-type string           The type of service to create for federation API server. Options: 'LoadBalancer' (default), 'NodePort'. (default "LoadBalancer")
      --apiserver-arg-overrides string           comma separated list of federation-apiserver arguments to override: Example "--arg1=value1,--arg2=value2..."
      --apiserver-enable-basic-auth              Enables HTTP Basic authentication for the federation-apiserver. Defaults to false.
      --apiserver-enable-token-auth              Enables token authentication for the federation-apiserver. Defaults to false.
      --controllermanager-arg-overrides string   comma separated list of federation-controller-manager arguments to override: Example "--arg1=value1,--arg2=value2..."
      --dns-provider string                      Dns provider to be used for this deployment.
      --dns-provider-config string               Config file path on local file system for configuring DNS provider.
      --dns-zone-name string                     DNS suffix for this federation. Federated Service DNS names are published with this suffix.
      --dry-run                                  dry run without sending commands to server.
      --etcd-image string                        Image to use for etcd server. (default "gcr.io/google_containers/etcd:3.0.17")
      --etcd-persistent-storage                  Use persistent volume for etcd. Defaults to 'true'. (default true)
      --etcd-pv-capacity string                  Size of persistent volume claim to be used for etcd. (default "10Gi")
      --etcd-pv-storage-class string             The storage class of the persistent volume claim used for etcd.   Must be provided if a default storage class is not enabled for the host cluster.
      --federation-system-namespace string       Namespace in the host cluster where the federation system components are installed (default "federation-system")
      --host-cluster-context string              Host cluster context
      --image string                             Image to use for federation API server and controller manager binaries. (default "gcr.io/google_containers/hyperkube-amd64:v0.0.0-master+$Format:%h$")
      --kubeconfig string                        Path to the kubeconfig file to use for CLI requests.
```

**What you expected to happen**:

Have `nodeSelector` for both kubernetes and etcd.

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by dixudx**
**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #50716

**Special notes for your reviewer**:
/assign @gyliu513 

**Release note**:

```release-note
support nodeSelector in kubefed init
```

### CODE REVIEW

File: federation/pkg/kubefed/init/init.go [hunk 1]
```diff
@@ -173,6 +175,7 @@ func (o *initFederationOptions) Bind(flags *pflag.FlagSet, defaultServerImage, d
 	flags.Int32Var(&o.apiServerNodePortPort, apiserverPortFlag, 0, "Preferred port to use for api server nodeport service (0 for random port assignment). Valid only if '"+apiserverServiceTypeFlag+"=NodePort'.")
 	flags.BoolVar(&o.apiServerEnableHTTPBasicAuth, "apiserver-enable-basic-auth", false, "Enables HTTP Basic authentication for the federation-apiserver. Defaults to false.")
 	flags.BoolVar(&o.apiServerEnableTokenAuth, "apiserver-enable-token-auth", false, "Enables token authentication for the federation-apiserver. Defaults to false.")
+	flags.StringVar(&o.nodeSelectorString, "nodeSelectorString", "", "comma separated list of nodeSelector arguments: Example \"arg1=value1,arg2=value2...\"")
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init.go]: I'd prefer that we use `node-selector` but not `nodeSelectorString` to keep consistent with other parameters.

File: federation/pkg/kubefed/init/init.go [hunk 2]
```diff
@@ -702,7 +709,7 @@ func createPVC(clientset client.Interface, namespace, svcName, federationName, e
 	return clientset.Core().PersistentVolumeClaims(namespace).Create(pvc)
 }
 
-func createAPIServer(clientset client.Interface, namespace, name, federationName, serverImage, etcdImage, advertiseAddress, credentialsName string, hasHTTPBasicAuthFile, hasTokenAuthFile bool, argOverrides map[string]string, pvc *api.PersistentVolumeClaim, dryRun bool) (*extensions.Deployment, error) {
+func createAPIServer(clientset client.Interface, namespace, name, federationName, serverImage, etcdImage, advertiseAddress, credentialsName string, hasHTTPBasicAuthFile, hasTokenAuthFile bool, argOverrides, nodeSelector map[string]string, pvc *api.PersistentVolumeClaim, dryRun bool) (*extensions.Deployment, error) {
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init.go]: How about set the new added parameter to the end? ditto for others.

File: federation/pkg/kubefed/init/init_test.go [hunk 3]
```diff
@@ -292,6 +294,9 @@ func TestInitFederation(t *testing.T) {
 		if tc.apiserverEnableTokenAuth {
 			cmd.Flags().Set("apiserver-enable-token-auth", "true")
 		}
+		if tc.nodeSelector != "" {
+			cmd.Flags().Set("nodeSelectorString", tc.nodeSelector)
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: Here should be updated to `node-selector` as well per my first comment.

File: federation/pkg/kubefed/init/init_test.go [hunk 4]
```diff
@@ -889,6 +894,11 @@ func fakeInitHostFactory(apiserverServiceType v1.ServiceType, federationName, na
 	sort.Strings(apiserverArgs)
 	apiserverCommand = append(apiserverCommand, apiserverArgs...)
 
+	nodeSelector, err := marshallOverrides(nodeSelectorString)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling --nodeSelectorString: %v", err)
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: here as well for `node-selector`

**Review by gyliu513 — COMMENTED** [review]
Thanks @dixudx , minor comments

**Comment by dixudx** [comment]
@gyliu513 Updated. PTAL. Thanks.

File: federation/pkg/kubefed/init/init_test.go [hunk 5]
```diff
@@ -889,6 +894,11 @@ func fakeInitHostFactory(apiserverServiceType v1.ServiceType, federationName, na
 	sort.Strings(apiserverArgs)
 	apiserverCommand = append(apiserverCommand, apiserverArgs...)
 
+	nodeSelector, err := marshallOverrides(nodeSelectorString)
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling --node-selector: %v", err)
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: here should be `t.Fatalf`

Comment (dixudx) [inline federation/pkg/kubefed/init/init_test.go]: @gyliu513 This is not a testcase, just a fake method. `t.Fatalf` is not working here.

Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: Ah, I see, thanks.

File: federation/pkg/kubefed/init/init_test.go [hunk 6]
```diff
@@ -621,7 +626,7 @@ func TestCertsHTTPS(t *testing.T) {
 	}
 }
 
-func fakeInitHostFactory(apiserverServiceType v1.ServiceType, federationName, namespaceName, advertiseAddress, lbIp, dnsZoneName, serverImage, etcdImage, dnsProvider, dnsProviderConfig, etcdPersistence, etcdPVCapacity, etcdPVStorageClass, apiserverOverrideArg, cmOverrideArg, tmpDirPath string, apiserverEnableHTTPBasicAuth, apiserverEnableTokenAuth, isRBACAPIAvailable bool) (cmdutil.Factory, error) {
+func fakeInitHostFactory(apiserverServiceType v1.ServiceType, federationName, namespaceName, advertiseAddress, lbIp, dnsZoneName, serverImage, etcdImage, dnsProvider, dnsProviderConfig, etcdPersistence, etcdPVCapacity, etcdPVStorageClass, apiserverOverrideArg, cmOverrideArg, nodeSelectorString, tmpDirPath string, apiserverEnableHTTPBasicAuth, apiserverEnableTokenAuth, isRBACAPIAvailable bool) (cmdutil.Factory, error) {
```
Comment (gyliu513) [inline federation/pkg/kubefed/init/init_test.go]: put the new parameter to the end?

Comment (dixudx) [inline federation/pkg/kubefed/init/init_test.go]: @gyliu513 Updated. PTAL.

**Review by gyliu513 — COMMENTED** [review]
/lgtm

**Comment by dixudx** [comment]
/assign @nikhiljindal @quinton-hoole
PTAL. Need your `/approval`. Thanks.

**Pull request merged**